src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs

```diff
@@ -1321,10 +1321,21 @@ private static JsonWebToken ValidateSignature(string token, TokenValidationParam
 
             }
 
+            // Get information on where keys used during token validation came from for debugging purposes.
+            var keysInTokenValidationParameters = TokenUtilities.GetAllSigningKeys(validationParameters);
+            var keysInConfiguration = TokenUtilities.GetAllSigningKeys(configuration);
+            var numKeysInTokenValidationParameters = keysInTokenValidationParameters.Count();
+            var numKeysInConfiguration = keysInConfiguration.Count();
+
             if (kidExists)
             {
                 if (kidMatched)
-                    throw LogHelper.LogExceptionMessage(new SecurityTokenInvalidSignatureException(LogHelper.FormatInvariant(TokenLogMessages.IDX10511, keysAttempted, jwtToken.Kid, exceptionStrings, jwtToken)));
+                {
+                    var isKidInTVP = keysInTokenValidationParameters.Any(x => x.KeyId.Equals(jwtToken.Kid, StringComparison.Ordinal));
+                    var keyLocation = isKidInTVP ? "TokenValidationParameters" : "Configuration";
+                    throw LogHelper.LogExceptionMessage(new SecurityTokenInvalidSignatureException(
+                        LogHelper.FormatInvariant(TokenLogMessages.IDX10511, keysAttempted, LogHelper.MarkAsNonPII(numKeysInTokenValidationParameters), LogHelper.MarkAsNonPII(numKeysInConfiguration), LogHelper.MarkAsNonPII(keyLocation), jwtToken.Kid, exceptionStrings, jwtToken)));
```

Comment:
for code with a large number of parameters, it reads easier if they are stacked, for example
```cs
 LogHelper.FormatInvariant(TokenLogMessages.IDX10511,
    keysAttempted, 
    LogHelper.MarkAsNonPII(numKeysInTokenValidationParameters), 
    LogHelper.MarkAsNonPII(numKeysInConfiguration),
    LogHelper.MarkAsNonPII(keyLocation), jwtToken.Kid, 
    exceptionStrings, 
    jwtToken)));

Reply:
I think that makes things a lot more readable. :) Thank you for the suggestion. Fixed! https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1786/commits/27f73791bb1e017be2e7bc1a0b4c25cba31b9a78